### PR TITLE
Re-instate correct eip1559 gas price calcualtions

### PR DESCRIPTION
**Pull request by piersy**
In this commit dfd1d38015d5ce7a174c84d4a4092308f668571a

Ponti mistakenly removed gas price calculation for eip1559 style transactions retreived via the RPC. This resulted in gasFeeCap being returned for all transactions except depost transactions.

So this commit fixes that but also adds support for calculating gas price for our migrated transactions as well.

Prior to the gingerbread hardfork celo stored the base fee in state as opposed to on the block, this PR does not add code to be able to retrieve the base fee from state, so for all pre-gingerbread eip1559 style transactions we have no way to calculate the gas price and therefore will return null.

Post gingerbread celo added the base fee to the block, this enables us to calculate the correct gas price for transactions using the native currency. For transactions using alternative fee currencies we still cannot calculate the correct gas price since we would need to look up the base fee and exchange rate from the state and we haven't added code to do this.

Post cel2 for celo eip1559 style transactions the baseFee in fee currency is stored on the receipt which means that we can calcualte the correct base fee for all transactions.

Fixes https://github.com/celo-org/celo-blockchain-planning/issues/382

### CODE REVIEW

File: core/types/receipt.go [hunk 1]
```diff
@@ -610,11 +610,17 @@ func (rs Receipts) DeriveFields(config *params.ChainConfig, hash common.Hash, nu
 		rs[i].Type = txs[i].Type()
 		rs[i].TxHash = txs[i].Hash()
 
-		// The post transition CeloDynamicFeeV2Txs set the baseFee in the receipt
-		if rs[i].BaseFee == nil {
-			rs[i].EffectiveGasPrice = txs[i].inner.effectiveGasPrice(new(big.Int), baseFee)
-		} else {
-			rs[i].EffectiveGasPrice = txs[i].inner.effectiveGasPrice(new(big.Int), rs[i].BaseFee)
+		// Pre-gingerbred the base fee was stored in state, but we don't try to recover it here, since A) we don't have
```
Comment (palango) [inline core/types/receipt.go:613]: Could we store it in the client somewhere. I guess it was usually constant for a range, so this wouldn't be horribly inefficient.

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -1564,21 +1564,42 @@ func newRPCTransaction(tx *types.Transaction, blockHash common.Hash, blockNumber
...
+
+		// Note that celo denominated txs always have the gas price denominated in celo (the native currency)
+		isNativeFeeCurrency := tx.FeeCurrency() == nil || tx.Type() == types.CeloDenominatedTxType
+		isGingerbread := config.IsGingerbread(new(big.Int).SetUint64(blockNumber))
+		isCel2 := config.IsCel2(blockTime)
+
+		if blockHash == (common.Hash{}) {
+			// This is a pending transaction, for pending transactions we set the gas price to gas fee cap.
+			result.GasPrice = (*hexutil.Big)(tx.GasFeeCap())
+		} else if isGingerbread && isNativeFeeCurrency && baseFee != nil {
```
Comment (palango) [inline internal/ethapi/api.go]: can `isGingerbread` and `baseFee != nil` ever happen?

File: internal/ethapi/celo_api_test.go [hunk 3]
```diff
@@ -115,63 +77,457 @@ func TestNewRPCTransactionLegacy(t *testing.T) {
...
-		assert.Nil(t, rpcTx.EthCompatible)
-		assert.Nil(t, rpcTx.MaxFeeInFeeCurrency)
+func TestNewRPCTransactionDynamicFee(t *testing.T) {
+	key, err := crypto.GenerateKey()
+	require.NoError(t, err)
+	feeCap := big.NewInt(1000)
+	tipCap := big.NewInt(100)
+
+	t.Run("PendingTransactions", func(t *testing.T) {
+		//For pending transactions we expect the gas price to the the gas fee cap.
```
Comment (palango) [inline internal/ethapi/celo_api_test.go]:
```suggestion
		// For pending transactions we expect the gas price to be the gas fee cap.
```

File: internal/ethapi/celo_api_test.go [hunk 4]
```diff
@@ -115,63 +77,457 @@ func TestNewRPCTransactionLegacy(t *testing.T) {
...
+			require.NoError(t, err)
+
+			rpcTx := newRPCTransaction(signed, blockhash, blockNumber, blockTime, transactionIndex, baseFee, config, receipt)
+			checkTxFields(t, signed, rpcTx, s, blockhash, blockNumber, transactionIndex, overrides)
+		})
 	})
 }
 
-func checkRPCTransactionFields(
+func allEnabledChainConfig() *params.ChainConfig {
```
Comment (palango) [inline internal/ethapi/celo_api_test.go:405]: Any specific readon why `AllDevChainProtocolChanges` isn't enough?

File: core/types/receipt.go (hunk 1 above)
Comment (piersy) [reply core/types/receipt.go:613]: We could hardcode it like the gas limit, but we could also fetch it from the block receipts, which might be a bit nicer. But I'd consider either to be outside the scope of this PR.

There's a ticket for this here - https://github.com/celo-org/celo-blockchain-planning/issues/384

File: internal/ethapi/api.go (hunk 2 above)
Comment (piersy) [reply internal/ethapi/api.go]: Do you mean. `isGingerbread` and `baseFee == nil`? Which sounds like it shouldn't happen, the `baseFee != nil` condition was taken from upstream, but thinking about it more it seems like that is to account for blocks occurring before eip1559. I will investigate more.

File: internal/ethapi/celo_api_test.go (hunk 4 above)
Comment (piersy) [reply internal/ethapi/celo_api_test.go:405]: I'm modifying the config in the tests, so I don't want those modifications affecting other tests.

File: internal/ethapi/api.go (hunk 2 above)
Comment (piersy) [reply internal/ethapi/api.go]: Yep, so I think that condition was to account for pre eip1559 blocks, the only case I can see where `baseFee` would be nil is for pending transactions which are handled by the above condition. So I think it is safe to remove the check on the base fee here.

**Review by palango — APPROVED** [review]

**Pull request merged**